doc/Messaging.md

```diff
@@ -0,0 +1,219 @@
+# Using the VRDR.Messaging Library
+
+This document describes how to use the VRDR.Messaging library to simplify implementation of VRDR message exchange flows.
+
+## Death Record Submission
+
+The diagram below illustrates the messages exchanged during submission of a death record and the subsequent return of coded causes of death and race and ethnicity information.
+
+![Message Exchange Pattern for Death Record Submission and Coding Response](submission.png)
+
+### Submit Death Record
+
+A vital records jurisdiction can create a death record submission as follows:
+
+```cs
+// Create a DeathRecord
+DeathRecord record = ...;
+
+// Create a submission message
+DeathRecordSubmission message = new DeathRecordSubmission(record);
+message.MessageSource = "https://example.com/juristdiction/message/endpoint";
+
+// Create a JSON representation of the message (XML is also supported via the ToXML method)
+string jsonMessage = message.ToJSON();
+
+// Send the JSON message
+...
+```
+
+The `DeathRecordSubmission` constructor will extract information from the supplied `DeathRecord` to populate the corresponding message property values (`StateIdentifier`, `CertificateNumber`, `NCHSIdentifier`) automatically.
+
+### Extract Death Record
+
+On receipt of a message, NCHS can parse it, determine the type of the message, and extract a death record using the following steps:
+
+```cs
+// Get the message as a stream
+StreamReader messageStream = ...;
+
+// Parse the message
+BaseMessage message = BaseMessage.Parse(messageStream);
+
+// Switch to determine the type of message
+switch(message)
+{
+    case DeathRecordSubmission submission:
+        var record = submission.DeathRecord;
+        var nchsId = submission.NCHSIDentifier;
+        ProcessSubmission(record, nchsId);
+        break;
+    ...
+}
+```
+
+### Acknowledge Death Record
+
+If extraction was succesful, NCHS can generate an acknowledgement message and send it to the submitting jurisdiction. An `AckMessage` constructor is available that accepts a source message parameter (`submission` in the example below) and this is used to initialize `AckMessage` properties:
+
+- `AckMessage.MessageDestination` will be assigned the value of `source.MessageSource`
+- `AckMessage.AckedMessageId` will be assigned the value of `source.MessageId`
+- `StateIdentifier` will be assigned the value of `source.StateIdentifier`
+- `CertificateNumber` will be assigned the value of `source.CertificateNumber`
+- `NCHSIdentifier` will be assigned the value of `source.NCHSIdentifier`
+
+The constructor will also default the value of the `MessageSource` property to `http://nchs.cdc.gov/vrdr_submission` which is suitable for messaging originating from NCHS.
```

Comment:
Wondering if this default makes sense given the jurisdiction use of Acks (as you accurately note below)... can we populate the MessageSource property from the MessageDestination of the message being acknowledged? My intuition is no because that inappropriately trusts the contents of message being acknowledged...

Reply:
Nice idea. On balance I think defaulting the ack source to the acked message destination is probably a better default that the current fixed one.